nixos/modules/services/networking/ssh/sshd.nix

```diff
@@ -145,6 +207,52 @@ in
         '';
       };
 
+      matches = mkOption {
+        default = [];
+        example = literalExample ''
+          [
+            {
+              match = {
+                "10.23.42.0/24" = "Address";
+                "somebody" = "User";
+              };
+              config = {
+                forwardX11 = true;
+                permitRootLogin = "yes";
+              };
+            }
+            {
+              match."10.23.42.0/24" = "Address";
+              config.logLevel = "ERROR";
+            }
+          ]
+        '';
+
+        description = ''
+          Declaratively specify matches like <literal>Match Address 10.23.42.0/24</literal>.
+        '';
+
+        type = types.listOf (types.submodule {
+          options = {
+            match = mkOption {
+              type = types.attrsOf (types.enum [ "Address" "User" ]);
+              description = ''
+                Configure matches for openssh.
+              '';
+
+              apply = x: concatStringsSep " " (mapAttrsToList (match: type: "${type} ${match}") x);
+            };
+
+            config = mkOption {
+              type = types.submodule ({ options = generalOptions; });
```

Comment:
We probably want to merge the values here with the configs from `services.openssh`. So if somebody sets `services.openssh.permitRootLogin` to `no`, each `Match` group should also have `permitRootLogin` set to `no`.

But I'll wait for some reviews from other folks, before spending more time on this :)